active/XamlServiceProvider/XamlServiceProviderApi.md

```diff
@@ -0,0 +1,272 @@
+
+    <Grid>
+        <TextBlock Text="{local:TestMarkupExtension}" />
+    </Grid>
+</Page>
+```
+
+## IUriContext 
+
+The following example shows a custom markup extension that converts a relative URI to an absolute URI. It does this by getting the IUriContext from the IXamlServiceProvider parameter passed in to ProvideValue. IUriContext provides the URI of the markup being loaded.
+
+This markup extension enables the following markup, where a property requires an absolute URI:
+
+```xml
+<SomeControl AbsoluteUri="{RelativeUriResolver 'image.jpg'}"
+```
+
+```cs
+public class RelativeUriResolver : MarkupExtension
+{
+    public RelativeUriResolver(Uri uri)
+    {
+        SourceUri = uri;
+    }
+    public Uri SourceUri { get; set; }
+    protected override object ProvideValue(IXamlServiceProvider provider)
+    {
+        if(SourceUri == null || SourceUri.IsAbsoluteUri)
+        {
+            return SourceUri;
+        }
+
+        var uriContext = provider.GetService(typeof(IUriContext)) as IUriContext;
+        if (uriContext != null || uriContext.BaseUri != null)
+        {
+            // For example this will calculate "ms-appx:///Controls/ColorPicker.xaml" + "colorImage.jpg" to 
+            // "ms-appx:///Controls/colorImage.jpg"
+            if (Uri.TryCreate(uriContext.BaseUri, SourceUri, out var absoluteUri))
+            {
+                return absoluteUri;
+            }
+        }
+
+        return SourceUri;
+    }
+}
+```
+
+# API Notes/Remarks
+
+## MarkupExtension.ProvideValue(IXamlServiceProvider) virtual method
+
+When implemented in a derived class, returns an object that is provided as the value of the target property for this markup extension.
+
+**Remarks**
+
+The IXamlServiceProvider argument can be used to retrieve the following services:
+* `IProvideValueTarget`, which provides information about the type/property the markup extension is being applied to. This aligns with WPF's [IProvideValueTarget](https://docs.microsoft.com/dotnet/api/System.Windows.Markup.IProvideValueTarget).
+* `IRootObjectProvider`, which provides a reference to the root object in the markup. This aligns with WPF's [IRootObjectProvider](https://docs.microsoft.com/dotnet/api/System.Xaml.IRootObjectProvider).
+* `IUriContext` , which provides the base URI of the markup. This aligns with WPF's [IUriContext](https://docs.microsoft.com/dotnet/api/System.Windows.Markup.IUriContext).
+* `IXamlTypeResolver`, which binds a Xaml markup name to a type. This aligns with WPF's [IXamlTypeResolver](https://docs.microsoft.com/en-us/dotnet/api/system.windows.markup.ixamltyperesolver).
+
+There are two overloads of the virtual ProvideValue method, with/without the IXamlServiceProvider parameter. During Xaml markup loading, ProvideValue(IXamlServiceProvider) is called. Its virtual implementatiohn calls ProvideValue(). So not overriding ProvideValue(IXamlServiceProvider), or calling the base implementation of it, calls ProvideValue().

```

Comment:
If my MarkupProvider overrides the `ProvideValue(IXamlServiceProvider)` but not the `ProvideValue()`, what happens? Are there ways that the 0-parameter could get called even if the 1-parameter one is present? Say, because some other app code grabs the MarkupProvider and starts calling ProvideValue directly?

Reply:
This paragraph tries to answer this: the virtual implementation of ProvideValue(IXamlServiceProvider) calls ProvideValue(), which returns null.